hoomd/hpmc/IntegratorHPMCMonoNEC.h

```diff
@@ -0,0 +1,1368 @@
+    provides the current value. The parameter *mode* controls whether the returned counts are absolute, relative
+    to the start of the run, or relative to the start of the last executed step.
+*/
+template< class Shape >
+hpmc_nec_counters_t IntegratorHPMCMonoNEC< Shape >::getNECCounters(unsigned int mode)
+    {
+    ArrayHandle<hpmc_nec_counters_t> h_nec_counters(m_nec_count_total, access_location::host, access_mode::read);
+    hpmc_nec_counters_t result;
+
+    if (mode == 0)
+        result = h_nec_counters.data[0];
+    else if (mode == 1)
+        result = h_nec_counters.data[0] - m_nec_count_run_start;
+    else
+        result = h_nec_counters.data[0] - m_nec_count_step_start;
+
+#ifdef ENABLE_MPI
+    if (this->m_comm)
+        {
+        // MPI Reduction to total result values on all nodes.
+        MPI_Allreduce(MPI_IN_PLACE, &result.chain_start_count,        1, MPI_LONG_LONG_INT, MPI_SUM, this->m_exec_conf->getMPICommunicator());
+        MPI_Allreduce(MPI_IN_PLACE, &result.chain_at_collision_count, 1, MPI_LONG_LONG_INT, MPI_SUM, this->m_exec_conf->getMPICommunicator());
+        MPI_Allreduce(MPI_IN_PLACE, &result.chain_no_collision_count, 1, MPI_LONG_LONG_INT, MPI_SUM, this->m_exec_conf->getMPICommunicator());
+        MPI_Allreduce(MPI_IN_PLACE, &result.distance_queries,         1, MPI_LONG_LONG_INT, MPI_SUM, this->m_exec_conf->getMPICommunicator());
+        MPI_Allreduce(MPI_IN_PLACE, &result.overlap_err_count,        1, MPI_UNSIGNED,      MPI_SUM, this->m_exec_conf->getMPICommunicator());
+        }
+#endif
+    return result;
+    }
+
+
+template< class Shape >
+void IntegratorHPMCMonoNEC< Shape >::update(uint64_t timestep)
+    {
+    this->m_exec_conf->msg->notice(10) << "HPMCMonoEC update: " << timestep << std::endl;
+    IntegratorHPMC::update(timestep);
+
+    // get needed vars
+    ArrayHandle<hpmc_counters_t> h_counters(this->m_count_total, access_location::host, access_mode::readwrite);
+    hpmc_counters_t& counters = h_counters.data[0];
+    // m_count_step_start = h_counters.data[0]; // in IntegratorHPMC
+
+    ArrayHandle<hpmc_nec_counters_t> h_nec_counters(m_nec_count_total, access_location::host, access_mode::readwrite);
+    hpmc_nec_counters_t& nec_counters = h_nec_counters.data[0];
+    m_nec_count_step_start = h_nec_counters.data[0];
+
+
+    const BoxDim& box = this->m_pdata->getBox();
+    unsigned int ndim = this->m_sysdef->getNDimensions();
+
+    #ifdef ENABLE_MPI
+    // compute the width of the active region
+    Scalar3 npd = box.getNearestPlaneDistance();
+    Scalar3 ghost_fraction = this->m_nominal_width / npd;
+
+    vec3<Scalar> lattice_x = vec3<Scalar>(box.getLatticeVector(0));
+    vec3<Scalar> lattice_y = vec3<Scalar>(box.getLatticeVector(1));
+    vec3<Scalar> lattice_z = vec3<Scalar>(box.getLatticeVector(2));
+
+    vec3<Scalar> normal_x = cross(lattice_y,lattice_z);
+    normal_x /= sqrt(dot(normal_x,normal_x));
```

Comment:
Please the sqrt functions written for Hoomd in `hoomd/HOOMDMath.h`
 ```suggestion
    normal_x *= fast::rsqrt(dot(normal_x,normal_x));
```